bundles/org.openhab.binding.intellicenter2/src/main/java/org/openhab/binding/intellicenter2/internal/protocol/ICProtocol.java

```diff
@@ -0,0 +1,288 @@
+    private final ConcurrentHashMap<String, List<NotifyListListener>> subscriptions = new ConcurrentHashMap<>();
+    private final ExecutorService listenersNotifier;
+
+    public ICProtocol(IntelliCenter2Configuration config) throws IOException {
+        this.config = config;
+        this.listenersNotifier = newSingleThreadExecutor(
+                new ThreadFactoryBuilder().setDaemon(true).setNameFormat("ic-protocol-listeners-%s").build());
+        connect();
+        this.systemInfo = querySystemInfo();
+    }
+
+    private void connect() throws IOException {
+        this.clientSocket.set(new Socket(config.hostname, config.port));
+        this.clientSocket.get().setKeepAlive(true);
+        this.out.set(new BufferedOutputStream(new DataOutputStream(clientSocket.get().getOutputStream())));
+        this.in.set(new BufferedReader(new InputStreamReader(clientSocket.get().getInputStream())));
+        this.writerService = newSingleThreadExecutor(
+                new ThreadFactoryBuilder().setDaemon(true).setNameFormat("ic-protocol-writer-%s").build());
+        this.readerService = newSingleThreadExecutor(
+                new ThreadFactoryBuilder().setDaemon(true).setNameFormat("ic-protocol-reader-%s").build());
+
+        this.readRunnable = new ICRead();
+        this.readerService.submit(readRunnable);
+    }
+
+    private void reconnect() throws IOException {
+        close();
+        connect();
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (readRunnable != null) {
+            readRunnable.stop = true;
+        }
+        safeClose(out.get());
+        safeClose(in.get());
+        safeClose(clientSocket.get());
+        try {
+            if (writerService != null) {
+                writerService.shutdownNow();
+            }
+            if (readerService != null) {
+                readerService.shutdownNow();
+            }
+        } catch (Exception ignored) {
+        }
+    }
+
+    private static void safeClose(@Nullable Closeable c) {
+        try {
+            if (c != null) {
+                c.close();
+            }
+        } catch (IOException ignored) {
+        }
+    }
+
+    public SystemInfo getSystemInfo() {
+        return systemInfo;
+    }
+
+    private SystemInfo querySystemInfo() {
+        return new SystemInfo(getUnchecked(submit(SystemInfo.REQUEST)).getObjectList().get(0));
+    }
+
+    public synchronized void subscribe(NotifyListListener listener, RequestObject request) {
+        subscriptions.putIfAbsent(request.getObjectName(), new CopyOnWriteArrayList<>());
+        subscriptions.get(request.getObjectName()).add(listener);
+        String response = getUnchecked(submit(ICRequest.requestParamList(request))).getResponse();
+        if (!"200".equals(response)) {
+            logger.error("Error subscribing listener {} for request {}", listener, request);
```

Comment:
I suppose this should not prevent the binding from working, so probably log level can be reduced?